File: src-server/src/sessions.rs
use serde::Serialize;
use serde_json::Value;
use std::{
    io::BufRead,
    path::{Path, PathBuf},
    time::UNIX_EPOCH,
};
use walkdir::WalkDir;

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSessions {
    pub cwd: String,
    pub name: String,
    pub sessions: Vec<SessionInfo>,
}

#[derive(Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SessionInfo {
    pub id: String,
    pub name: Option<String>,
    pub path: String,
    pub cwd: Option<String>,
    pub created_at: u64,
    pub modified_at: u64,
    pub message_count: usize,
    pub tokens: u64,
    pub cost: f64,
}

pub fn sessions_root() -> Result<PathBuf, String> {
    if let Ok(base) = std::env::var("PI_CODING_AGENT_DIR") {
        return Ok(PathBuf::from(base).join("sessions"));
    }
    dirs::home_dir()
        .map(|home| home.join(".pi/agent/sessions"))
        .ok_or("Unable to find the Pi sessions directory".into())
}

fn timestamp(metadata: &std::fs::Metadata, created: bool) -> u64 {
    let value = if created {
        metadata.created().or_else(|_| metadata.modified())
    } else {
        metadata.modified()
    };
    value
        .ok()
        .and_then(|time| time.duration_since(UNIX_EPOCH).ok())
        .map(|v| v.as_secs())
        .unwrap_or(0)
}

fn parse_session(path: &Path) -> Result<SessionInfo, String> {
    let metadata = path.metadata().map_err(|error| error.to_string())?;
    if metadata.len() > 50 * 1024 * 1024 {
        return Err("Session file exceeds the indexing limit".into());
    }
    let file = std::fs::File::open(path).map_err(|error| error.to_string())?;
    let mut id = path
        .file_stem()
        .and_then(|v| v.to_str())
        .unwrap_or("session")
        .to_string();
    let mut name = None;
    let mut cwd = None;
    let mut message_count = 0;
    let mut tokens = 0;
    let mut cost = 0.0;
    for line in std::io::BufReader::new(file).lines().map_while(Result::ok) {
        let Ok(value) = serde_json::from_str::<Value>(&line) else {
            continue;
        };
        if let Some(value_id) = value.get("id").and_then(Value::as_str) {
            if value.get("type").and_then(Value::as_str) == Some("session") {
                id = value_id.to_string();
            }
        }
        if name.is_none() {
            name = value
                .get("name")
                .or_else(|| value.get("sessionName"))
                .and_then(Value::as_str)
                .map(str::to_string);
        }
        if cwd.is_none() {
            cwd = value.get("cwd").and_then(Value::as_str).map(str::to_string);
        }
        if value.get("type").and_then(Value::as_str) == Some("message")
            || value.get("role").is_some()
        {
            message_count += 1;
        }
        let usage = value
            .pointer("/message/usage")
            .or_else(|| value.get("usage"));
        if let Some(usage) = usage {
            tokens += usage
                .get("totalTokens")
                .or_else(|| usage.get("total_tokens"))
                .and_then(Value::as_u64)
                .unwrap_or(0);
            cost += usage
                .pointer("/cost/total")
                .or_else(|| usage.get("cost"))
                .and_then(Value::as_f64)
                .unwrap_or(0.0);
        }
    }
    Ok(SessionInfo {
        id,
        name,
        path: path.to_string_lossy().into_owned(),
        cwd,
        created_at: timestamp(&metadata, true),
        modified_at: timestamp(&metadata, false),
        message_count,
        tokens,
        cost,
    })
}

pub fn list_all_sessions_grouped() -> Result<Vec<ProjectSessions>, String> {
    let root = sessions_root()?;
    if !root.exists() {
        return Ok(Vec::new());
    }
    let mut sessions = WalkDir::new(root)
        .follow_links(false)
        .max_depth(4)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|entry| {
            entry.file_type().is_file()
                && entry.path().extension().and_then(|v| v.to_str()) == Some("jsonl")
        })
        .take(10_000)
        .filter_map(|entry| parse_session(entry.path()).ok())
        .collect::<Vec<_>>();
    sessions.sort_by_key(|session| std::cmp::Reverse(session.modified_at));
    let mut groups: std::collections::BTreeMap<String, Vec<SessionInfo>> =
        std::collections::BTreeMap::new();
    for session in sessions {
        let cwd = session.cwd.clone().unwrap_or_else(|| String::from("unknown"));
        groups.entry(cwd).or_default().push(session);
    }
    let mut projects: Vec<ProjectSessions> = groups
        .into_iter()
        .map(|(cwd, mut sessions)| {
            sessions.sort_by_key(|s| std::cmp::Reverse(s.modified_at));
            let name = std::path::Path::new(&cwd)
                .file_name()
                .and_then(|v| v.to_str())
                .unwrap_or(&cwd)
                .to_string();
            ProjectSessions { cwd, name, sessions }
        })
        .collect();
    projects.sort_by_key(|p| {
        std::cmp::Reverse(p.sessions.first().map(|s| s.modified_at).unwrap_or(0))
    });
    Ok(projects)
}

pub fn list_sessions() -> Result<Vec<SessionInfo>, String> {
    let root = sessions_root()?;
    if !root.exists() {
        return Ok(Vec::new());
    }
    let mut sessions = WalkDir::new(root)
        .follow_links(false)
        .max_depth(4)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|entry| {
            entry.file_type().is_file()
                && entry.path().extension().and_then(|v| v.to_str()) == Some("jsonl")
        })
        .take(10_000)
        .filter_map(|entry| parse_session(entry.path()).ok())
        .collect::<Vec<_>>();
    sessions.sort_by_key(|session| std::cmp::Reverse(session.modified_at));
    Ok(sessions)
}

pub fn delete_session(session_path: String) -> Result<(), String> {
    let root = std::fs::canonicalize(sessions_root()?).map_err(|error| error.to_string())?;
    let target = std::fs::canonicalize(&session_path).map_err(|error| error.to_string())?;
    if !target.starts_with(root)
        || target.extension().and_then(|value| value.to_str()) != Some("jsonl")
    {
        return Err("Session path is outside the Pi sessions directory".into());
    }
    std::fs::remove_file(target).map_err(|error| format!("Unable to delete session: {error}"))
}

File: src-server/sessions.ts
import { readdirSync, statSync, readFileSync, existsSync, mkdirSync, unlinkSync } from "node:fs"
import { join, basename, resolve, normalize } from "node:path"
import { homedir } from "node:os"

export interface ProjectSessions {
  cwd: string
  name: string
  available: boolean
  sessions: SessionInfo[]
}

export interface SessionInfo {
  id: string
  name?: string
  path: string
  cwd?: string
  createdAt: number
  modifiedAt: number
  messageCount: number
  tokens: number
  cost: number
}

// ── Path helpers (matches Pi's implementation in session-manager.ts) ──

function sessionsRoot(): string {
  if (process.env.PI_CODING_AGENT_DIR) {
    return join(process.env.PI_CODING_AGENT_DIR, "sessions")
  }
  return join(homedir(), ".pi/agent/sessions")
}

/** Encode a cwd into a directory name — exactly how Pi does it. */
function encodeCwd(cwd: string): string {
  const resolved = resolve(normalize(cwd))
  return `--${resolved.replace(/^[/\\]/, "").replace(/[/\\:]/g, "-")}--`
}

/** Read just the first line of a session file to extract the header. */
function readSessionHeader(filePath: string): Record<string, unknown> | null {
  try {
    const fd = readFileSync(filePath, "utf8")
    const nl = fd.indexOf("\n")
    const firstLine = nl >= 0 ? fd.slice(0, nl) : fd
    if (!firstLine) return null
    const header = JSON.parse(firstLine)
    if (header?.type !== "session" || typeof header.id !== "string") return null
    return header
  } catch {
    return null
  }
}

/** Fast scan: read header line only, count messages by scanning the file. */
function parseSession(filePath: string): SessionInfo | undefined {
  const stat = statSafe(filePath)
  if (!stat || stat.size > 50 * 1024 * 1024) return undefined

  const header = readSessionHeader(filePath)
  if (!header) return undefined

  const id = String(header.id)
  // Pi stores cwd in the session header — this is the canonical source.
  const cwd = typeof header.cwd === "string" ? header.cwd : undefined
  // Name may be on the header itself (older Pi versions).
  let name: string | undefined = typeof header.name === "string" ? header.name : undefined

  // Scan the file for message count, tokens, cost, and name.
  let messageCount = 0
  let tokens = 0
  let cost = 0

  const content = readFileSync(filePath, "utf8")
  const lines = content.split("\n")
  let firstLine = true
  for (const rawLine of lines) {
    if (firstLine) { firstLine = false; continue }
    if (!rawLine.trim()) continue
    let value: any
    try { value = JSON.parse(rawLine) } catch { continue }

    if (name === undefined) {
      name = value.sessionName ?? value.name ?? undefined
    }
    if (value.type === "session_info" && value.name) {
      name = value.name.trim()
    }
    if (value.type === "message" || value.role !== undefined) {
      messageCount += 1
    }
    const usage = value.message?.usage ?? value.usage
    if (usage) {
      tokens += usage.totalTokens ?? usage.total_tokens ?? 0
      cost += usage.cost?.total ?? usage.cost ?? 0
    }
  }

  return {
    id,
    name,
    path: filePath,
    cwd,
    createdAt: Math.floor((stat.ctimeMs || stat.mtimeMs || 0) / 1000),
    modifiedAt: Math.floor((stat.mtimeMs || 0) / 1000),
    messageCount,
    tokens,
    cost,
  }
}

function statSafe(path: string) {
  try { return statSync(path) } catch { return undefined }
}

// ── Public API ────────────────────────────────────────────────────────

/** List all sessions grouped by cwd, scanning every directory under the
 *  Pi sessions root — exactly like Pi's SessionManager.listAll(). */
export function listAllSessionsGrouped(): ProjectSessions[] {
  const root = sessionsRoot()
  if (!existsSync(root)) return []

  const groups = new Map<string, SessionInfo[]>()

  try {
    const entries = readdirSync(root, { withFileTypes: true })
    for (const entry of entries) {
      if (!entry.isDirectory()) continue
      const dir = join(root, entry.name)
      let files: string[]
      try { files = readdirSync(dir) } catch { continue }
      for (const file of files) {
        if (!file.endsWith(".jsonl")) continue
        const session = parseSession(join(dir, file))
        if (!session) continue
        const cwd = session.cwd ?? ""
        if (!groups.has(cwd)) groups.set(cwd, [])
        groups.get(cwd)!.push(session)
      }
    }
  } catch {
    return []
  }

  const projects: ProjectSessions[] = []
  for (const [cwd, sessions] of groups) {
    sessions.sort((a, b) => b.modifiedAt - a.modifiedAt)
    const available = cwd !== "" && statSafe(cwd)?.isDirectory() === true
    const name = cwd === "" ? "Unknown" : basename(cwd) || cwd
    projects.push({ cwd, name, available, sessions })
  }
  // Sort: available first (by recency), then unavailable (also by recency).
  projects.sort((a, b) => {
    if (a.available !== b.available) return a.available ? -1 : 1
    return (b.sessions[0]?.modifiedAt ?? 0) - (a.sessions[0]?.modifiedAt ?? 0)
  })
  return projects
}

export function listSessions(): SessionInfo[] {
  const groups = listAllSessionsGrouped()
  return groups.flatMap((g) => g.sessions).sort((a, b) => b.modifiedAt - a.modifiedAt)
}

export function deleteSession(sessionPath: string): void {
  const root = sessionsRoot()
  if (!existsSync(root)) throw new Error("Sessions directory unavailable")
  const resolvedRoot = resolve(normalize(root))
  const resolvedTarget = resolve(normalize(sessionPath))
  if (!resolvedTarget.startsWith(resolvedRoot)) {
    throw new Error("Session path is outside the Pi sessions directory")
  }
  if (!resolvedTarget.endsWith(".jsonl")) {
    throw new Error("Session path must point to a .jsonl file")
  }
  try { unlinkSync(resolvedTarget) } catch (e: any) {
    // ENOENT means the file was already removed (e.g. by Pi compaction
    // or session replacement during switch_session). Treat as success.
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') return
    throw new Error(`Unable to delete session: ${e.message ?? e}`)
  }
}

/** Ensure the session directory for a cwd exists — Pi will create the
 *  actual session file when the first message is sent. */
export function ensureSessionDir(cwd: string): string {
  const root = sessionsRoot()
  const dirName = encodeCwd(cwd)
  const dir = join(root, dirName)
  if (!existsSync(dir)) {
    mkdirSync(dir, { recursive: true })
  }
  return dir
}
